Source/WebCore/css/StyleProperties.cpp

```diff
@@ -581,180 +581,211 @@ String StyleProperties::fontSynthesisValue() const
+        return makeString(top, ' ', right, ' ', bottom, ' ', left);
+    if (bottom != top)
+        return makeString(top, ' ', right, ' ', bottom);
+    if (right != top)
+        return makeString(top, ' ', right);
+    return top;
+}
 
-    auto top = propertyAt(findPropertyIndex(shorthand.properties()[0]));
-    auto right = propertyAt(findPropertyIndex(shorthand.properties()[1]));
-    auto bottom = propertyAt(findPropertyIndex(shorthand.properties()[2]));
-    auto left = propertyAt(findPropertyIndex(shorthand.properties()[3]));
-
-    if (top.isInherited() && right.isInherited() && bottom.isInherited() && left.isInherited())
-        return nameString(CSSValueInherit);
-
-    if (top.value()->isInitialValue() || right.value()->isInitialValue() || bottom.value()->isInitialValue() || left.value()->isInitialValue()) {
-        if (top.value()->isInitialValue() && right.value()->isInitialValue() && bottom.value()->isInitialValue() && left.value()->isInitialValue() && !top.isImplicit()) {
-            // All components are "initial" and "top" is not implicit.
-            return nameString(CSSValueInitial);
-        }
-        return String();
-    }
-
-    bool showLeft = !right.value()->equals(*left.value());
-    bool showBottom = !top.value()->equals(*bottom.value()) || showLeft;
-    bool showRight = !top.value()->equals(*right.value()) || showBottom;
+static bool isValueID(const RefPtr<CSSValue>& value)
+{
+    auto primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value.get());
+    return primitiveValue && primitiveValue->isValueID();
+}
 
-    StringBuilder result;
-    result.append(top.value()->cssText());
-    if (showRight) {
-        result.append(' ');
-        result.append(right.value()->cssText());
-    }
-    if (showBottom) {
-        result.append(' ');
-        result.append(bottom.value()->cssText());
-    }
-    if (showLeft) {
-        result.append(' ');
-        result.append(left.value()->cssText());
-    }
-    return result.toString();
+static bool isPair(const RefPtr<CSSValue>& value)
+{
+    auto primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value.get());
+    return primitiveValue && primitiveValue->isPair();
 }
 
 String StyleProperties::getLayeredShorthandValue(const StylePropertyShorthand& shorthand) const
 {
-    StringBuilder result;
-
-    const unsigned size = shorthand.length();
-    Vector<RefPtr<CSSValue>> values(size);
-    size_t numLayers = 0;
+    ASSERT(shorthand.length() <= 9);
 
-    for (unsigned i = 0; i < size; ++i) {
-        values[i] = getPropertyCSSValue(shorthand.properties()[i]);
-        if (values[i]->isBaseValueList())
-            numLayers = std::max(downcast<CSSValueList>(*values[i]).length(), numLayers);
-        else
-            numLayers = std::max<size_t>(1U, numLayers);
+    unsigned length = shorthand.length();
+    Vector<Ref<CSSValue>> values;
+    values.reserveInitialCapacity(length);
+    size_t numLayers = 1;
+    for (unsigned i = 0; i < length; ++i) {
```

Comment:
I think this loop can be a range-based loop as `I` is only used in the one place.